Added new hybrid search implementation

ccjoel opened the pull request:
New implementation does not use fuzzy matching on the keyword match side, as well as consolidates results of both keyword and semantic on the python server, by alternating results 3 at a time.

How it looks integrated with the UI:
![Screenshot from 2023-06-23 16-18-14](https://github.com/jataware/dojo/assets/1967061/c972815d-2ac8-463d-b1b8-9f726fe3af09)

Keyword search usually has higher match % than semantic. Does the back and forth between colors and match% bar sizes bother anyone else? See screenshot above. Got feedback on this and punted for later, as this dojo interface is not targeted for external users, but for debugging/tweaking search purposes.

hartleybrody commented on api/src/dojo.py:
```diff
@@ -139,35 +168,43 @@ def get_configs(model_id: str) -> List[DojoSchema.ModelConfig]:
+        params_to_delete = []
+        for param in params:
+            if param.get("template", {}).get("path") == path:
+                # TODO: see if this same param exists in other configs or directives?
```
I left this as a TODO since we haven't really figured out what the implementation would look like for https://github.com/jataware/shorthand/issues/24 in terms of what metadata we'd store with parameters.

hartleybrody commented on api/src/dojo.py:
```diff
@@ -139,35 +168,45 @@ def get_configs(model_id: str) -> List[DojoSchema.ModelConfig]:
+
+        modify_model(config["model_id"], ModelSchema.ModelMetadataPatchSchema(parameters=params))
+
+        # TODO remove s3_url and s3_url_raw from s3?
```
I didn't implement this here, since that would require adding `boto` as a dependency as well as needing to add a few env variables (`S3_ACCESS_KEY_ID`, etc) and I wanted to discuss if it made sense to keep that s3 logic centralized in shorthand instead of doing some of it in dojo and some of it in shorthand.

hartleybrody commented:
I bundled in some similar updates to address issue #109 as well, this is now ready for review.

mattprintz reviewed (approved):
Confirmed that the issue that I solved by adding the IDs was not recreated when you removed the IDs, so I think we're all good. Thanks.

mattprintz commented:
Oh, and the concern about removing the optional id from the pydantic model is moot because the model has the `extra = "allow"` option set so we should be good with that too.

hartleybrody commented:
> Oh, and the concern about removing the optional id from the pydantic model is moot because the model has the `extra = "allow"` option set so we should be good with that too.

Yeah, it took me a while to figure this out, but pydantic kept adding an `id` key to the JSON that was being submitted in the request, so it kept ending up with an empty `id` key in elasticsearch which seemed confusing. I eventually discovered that removing that line from the pydantic definition stopped it from adding that empty key.

ccjoel committed:
Added new hybrid search implementation that does not use fuzzy matching on the
keyword match side, as well as consolidates results of both keyword and semantic
on the python server, by alternating results 3 at a time.

ccjoel commented on api/scripts/hybrid_search.py:
```diff
@@ -42,7 +42,9 @@
     semantic_search_query,
     hybrid_query_v1,
     keyword_query_v1,
-    hybrid_query_v0
```
This is the query-comparison-report helper script. Feel free to ignore.

ccjoel commented on api/src/feature_queries.py:
```diff
@@ -1,5 +1,36 @@
+                "should": [
+                    {
+                        "match_phrase": {
+                            "name": phrase
```
New keyword query shared on slack. Without fuzzy matching. Used together with our semantic search implementation for the latest hybrid search.

ccjoel commented on ui/client/components/ViewFeatures.js:
```diff
@@ -364,7 +364,6 @@ const ViewFeatures = withStyles(() => ({
           LoadingOverlay: CustomLoadingOverlay
         }}
         loading={featuresLoading}
-        getRowId={(row) => `${row.owner_dataset.id}-${row.name}`}
```
This was an old addition, not necessary anymore as the features include an id, which DataGrid uses by default. The existing feature IDs are exactly this (a concatenation of their owner dataset and their name).

ccjoel committed:
Adding a param to a fn call, which was missing and was only working because the
missing value was the same as the default.

ccjoel committed:
Changed strategy on alternating matched queries for hybrid feature search.

ccjoel committed:
Adding hybrid search v2 (latest) to hybrid_search report/contrast script.

ccjoel committed:
Misc cleanup while reviewing PR.
